### PR TITLE
GH-39398: [C++][Parquet] DNM: benchmark for readLevels

### LINKED ISSUE #39398: [C++] [Parquet] Use std::count in parquet ColumnReader
### Describe the enhancement requested

I've found that for-loop here
https://github.com/apache/arrow/blob/7c3480e2f028f5881242f227f42155cf833efee7/cpp/src/parquet/column_reader.cc#L1055-L1073
transforms into

0xc0c2f0 <ReadLevels()+96>      inc    %rdx
0xc0c2f3 <ReadLevels()+99>      cmp    %rax,%rdx
0xc0c2f6 <ReadLevels()+102>     jge    0xc0c30c <ReadLevels()+124>
0xc0c2f8 <ReadLevels()+104>     cmp    %cx,(%r14,%rdx,2)
0xc0c2fd <ReadLevels()+109>     jne    0xc0c2f0 <ReadLevels()+96>
0xc0c2ff <ReadLevels()+111>     incq   0x0(%rbp)                                                   
0xc0c303 <ReadLevels()+115>     mov    (%rbx),%rax
0xc0c306 <ReadLevels()+118>     jmp    0xc0c2f0 <ReadLevels()+96>

That means that it uses iteration element by element and changes reference with incq
I think that the reason is that values_to_read and num_def_levels are not set as restrict. So the compiler can not optimize this to a more efficient way(for example using simd)

On my flamegraph this part showed ~10% of time spent

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Are there any user-facing changes?

* Closes: #39398

### CODE REVIEW

**Comment by mapleFU** [comment]
Please not merge this patch, this is just for benchmark

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 1]
```diff
@@ -141,6 +144,35 @@ static void ColumnReaderReadBatchInt32(::benchmark::State& state) {
...
+  BenchmarkHelper helper(repetition, /*num_pages=*/1, /*levels_per_page=*/16 * 80000);
+
+  // Vectors to read the values into.
+  std::vector<int32_t> read_values(batch_size, -1);
+  std::vector<int16_t> read_defs(batch_size, -1);
+  std::vector<int16_t> read_reps(batch_size, -1);
+  for (auto _ : state) {
+    state.PauseTiming();
+    Int32Reader* reader = helper.ResetColumnReader();
+    [[maybe_unused]] bool v = reader->HasNext();
```
Comment (mapleFU) [inline cpp/src/parquet/column_reader_benchmark.cc:163]: Using hasNext to trigger initialization

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 2]
```diff
@@ -141,6 +144,35 @@ static void ColumnReaderReadBatchInt32(::benchmark::State& state) {
...
 }
 
+// Benchmarks ReadBatch for ColumnReader with the following parameters in order:
+// - repetition: 0 for REQUIRED, 1 for OPTIONAL, 2 for REPEATED.
+// - batch_size: sets how many values to read at each call.
+static void ColumnReaderReadLevels(::benchmark::State& state) {
+  const auto repetition = static_cast<Repetition::type>(state.range(0));
+  const auto batch_size = static_cast<int64_t>(state.range(1));
+
+  BenchmarkHelper helper(repetition, /*num_pages=*/1, /*levels_per_page=*/16 * 80000);
```
Comment (mapleFU) [inline cpp/src/parquet/column_reader_benchmark.cc:154]: Using one page to make it simple

File: cpp/src/parquet/column_reader.h [hunk 3]
```diff
@@ -300,6 +300,9 @@ class TypedColumnReader : public ColumnReader {
                                           int16_t* rep_levels, int32_t* indices,
                                           int64_t* indices_read, const T** dict,
                                           int32_t* dict_len) = 0;
+
+  virtual void ReadLevels(int64_t batch_size, int16_t* def_levels, int16_t* rep_levels,
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader.h:304]: What do you think about making this private with a [Test Peer](https://abseil.io/tips/135) that can be used in the benchmark, so we can check this PR in?

**Comment by emkornfield** [comment]
Thank you @mapleFU it would be great to check this in (without the count changes) so we have a good benchmark for ReadLevels.  See comment on how to maybe do it without breaking abstractions.

**Comment by emkornfield** [comment]
I guess looking at CI it would take a little more work than this proof of concept to check it in.

**Comment by mapleFU** [comment]
It's just a quick poc for ReadLevels optimization... I think exporting it is so hacking, because `ReadLevels` is just "read levels in current page". So, `HasNext` is called, and I only maintaining one page...The interface would be weird here

**Comment by pitrou** [comment]
@ursabot please benchmark

**Comment by mapleFU** [comment]
Emm would regression benchmark un-related..?

**Comment by pitrou** [comment]
> Emm would regression benchmark un-related..?

None of them are related to Parquet.

**Pull request closed without merging**